### PR TITLE
Improve performance of filter_dict

### LINKED ISSUE #2062: Improve performance of filter_dict
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

When profiling a project using arrow-rs, found that filter_dict is among one of heavy stack trace:

```
arrow::compute::kernels::filter::filter_array::h7b1777b4117bb70a	
arrow::compute::kernels::filter::filter_dict::h72d7e1e318536f9f	
_$LT$arrow..array..array_dictionary..DictionaryArray$LT$T$GT$$u20$as$u20$core..convert..From$LT$arrow..array..data..ArrayData$GT$$GT$::from::h3716ed969de82ad7	
arrow::array::array::make_array::h48e79b04d13b81ef
```

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

<!---
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

Closes #2062.

# Rationale for this change
 
 <!---
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.
-->

```
filter context string dictionary (kept 1/2)                                                                                                                                                                                            
                        time:   [21.995 us 22.197 us 22.408 us]                                                                                                                                                                        
                        change: [-2.3630% -1.2909% -0.2122%] (p = 0.02 < 0.05)                                                                                                                                                         
                        Change within noise threshold.                                                                                                                                                                                 
Found 8 outliers among 100 measurements (8.00%)                                                                                                                                                                                        
  8 (8.00%) high mild                                                                                                                                                                                                                  
                                                                                                                                                                                                                                       
filter context string dictionary high selectivity (kept 1023/1024)                                                                                                                                                                     
                        time:   [10.567 us 10.650 us 10.735 us]                                                                                                                                                                        
                        change: [-3.2419% -2.2248% -1.2276%] (p = 0.00 < 0.05)
                        Performance has improved.

filter context string dictionary low selectivity (kept 1/1024)                                                                              
                        time:   [910.91 ns 917.49 ns 924.83 ns]
                        change: [-23.778% -22.819% -21.986%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe

filter context string dictionary w NULLs (kept 1/2)                                                                             
                        time:   [53.386 us 53.870 us 54.349 us]
                        change: [-5.8932% -4.8921% -3.8850%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

filter context string dictionary w NULLs high selectivity (kept 1023/1024)                                                                              
                        time:   [20.187 us 20.381 us 20.576 us]
                        change: [-2.2931% -1.2510% -0.2547%] (p = 0.02 < 0.05)
                        Change within noise threshold.
Found 6 outliers among 100 measurements (6.00%)
  5 (5.00%) high mild
  1 (1.00%) high severe

filter context string dictionary w NULLs low selectivity (kept 1/1024)                                                                              
                        time:   [1.2670 us 1.2742 us 1.2816 us]
                        change: [-24.367% -23.820% -23.263%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  4 (4.00%) high mild
```

# What changes are included in this PR?

<!---
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

# Are there any user-facing changes?


<!---
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!---
If there are any breaking changes to public APIs, please add the `breaking change` label.
-->

### CODE REVIEW

**Comment by viirya** [comment]
cc @sunchao

File: arrow/src/array/array_dictionary.rs [hunk 1]
```diff
@@ -151,6 +151,28 @@ impl<K: ArrowPrimitiveType> DictionaryArray<K> {
...
 
+    /// Create a new DictionaryArray directly from specified keys
+    /// (indexes into the dictionary) and values (dictionary)
+    /// array, and the corresponding ArrayData. This is used internally
+    /// for the usage like filter kernel.
+    ///
+    /// # Safety
+    ///
+    /// The input keys, values and data must form a valid DictionaryArray,
+    /// or undefined behavior can results.
```
Comment (sunchao) [inline arrow/src/array/array_dictionary.rs:162]: "can results" -> "can occur".

**Review by sunchao — APPROVED** [review]
LGTM

**Comment by viirya** [comment]
Thanks @sunchao

**Pull request merged**